src/PowerShellEditorServices/Extensions/Api/DocumentSymbolService.cs

```diff
@@ -0,0 +1,275 @@
+        /// <param name="scriptExtent">The script extent of the symbol</param>
+        /// <param name="filePath">The file path of the symbol</param>
+        /// <param name="sourceLine">The line contents of the given symbol (defaults to empty string)</param>
+        public SymbolReference(SymbolType symbolType, IScriptExtent scriptExtent)
+            : this(symbolType, scriptExtent.Text, scriptExtent)
+        {
+        }
+
+        /// <summary>
+        /// Constructs and instance of a SymbolReference
+        /// </summary>
+        /// <param name="symbolType">The higher level type of the symbol</param>
+        /// <param name="symbolName">The name of the symbol</param>
+        /// <param name="scriptExtent">The script extent of the symbol</param>
+        /// <param name="filePath">The file path of the symbol</param>
+        /// <param name="sourceLine">The line contents of the given symbol (defaults to empty string)</param>
+        public SymbolReference(
+            SymbolType symbolType,
+            string symbolName,
+            IScriptExtent scriptExtent)
+            : this(symbolType, scriptExtent, symbolName, filePath: string.Empty, sourceLine: string.Empty)
+        {
+        }
+
+        public SymbolReference(
+            SymbolType symbolType,
+            IScriptExtent scriptExtent,
+            string symbolName,
+            string filePath)
+            : this(symbolType, scriptExtent, symbolName, filePath, sourceLine: string.Empty)
+        {
+        }
+
+        public SymbolReference(SymbolType symbolType, IScriptExtent scriptExtent, string symbolName, string filePath, string sourceLine)
+        {
+            // TODO: Verify params
+            SymbolType = symbolType;
+            ScriptRegion = ScriptRegion.Create(scriptExtent);
+            SymbolName = symbolName;
+            FilePath = filePath;
+            SourceLine = sourceLine;
+
+            // TODO: Make sure end column number usage is correct
+        }
+
+        #region Properties
+
+        /// <summary>
+        /// Gets the symbol's type
+        /// </summary>
+        public SymbolType SymbolType { get; }
+
+        /// <summary>
+        /// Gets the name of the symbol
+        /// </summary>
+        public string SymbolName { get; }
+
+        /// <summary>
+        /// Gets the script extent of the symbol
+        /// </summary>
+        public ScriptRegion ScriptRegion { get; }
+
+        /// <summary>
+        /// Gets the contents of the line the given symbol is on
+        /// </summary>
+        public string SourceLine { get; }
+
+        /// <summary>
+        /// Gets the path of the file in which the symbol was found.
+        /// </summary>
+        public string FilePath { get; internal set; }
+
+        #endregion
+    }
+
+    /// <summary>
+    /// Service for registration of document symbol providers in PSES.
+    /// </summary>
+    public interface IDocumentSymbolService
```

Comment:
I've ported this out and made it available for registration, but given that CodeLens can't come out yet (and turns out we need to change that API anyway since VSCode has changed theirs) and this is slightly hacked together, maybe it's worth disabling compilation of this file for this release?